### PR TITLE
Config files in YAML format.

### PR DESCRIPTION
**Pull request by davidwyld**
@fredblain It would be great to know what you think of these changes to the config format.

At present, there is one yml file for the training, named eg `configs/config-sent-BiRNN.yml` and one for the general model, named `configs/default-config-BiRNN.yml`.

Before training, the default config is loaded into the `parameters` dictionary, then the specific config is loaded, and the user can overwrite any of the default options via this file. Also, as it is, the parameter key names are the same as before, so any that were dependent on others, ie `DATA_ROOT_PATH` was set based on `TASK_NAME`, are just set after the file is loaded.

Currently, this is hard-coded for proof-of-principle, but we can have this file specified somewhere by the user, or via choice of model or something. Overall, we can improve how this is handled later on.

### CODE REVIEW

File: predict.py [hunk 1]
```diff
@@ -30,12 +32,10 @@ def parse_args():
...
         help="model file (.h5) to use")
-    parser.add_argument("-c", "--config", required=False, help="Config pkl for loading the model configuration. "
-                                                               "If not specified, hyperparameters "
-                                                               "are read from config.py")
-    parser.add_argument("--save_path", required=False,
-        help="Directory path to save predictions to. "
-             "If not specified, defaults to STORE_PATH")
+    parser.add_argument("--save_path", required=False, help="Directory path to save predictions to. "
+                                                                "If not specified, defaults to STORE_PATH")
+    parser.add_argument("-es", "--evalset", required=False, help="Set to evaluate on."
```
Comment (drj11) [inline predict.py]: missing a space, `" "`, inside the string, after `on.`

File: predict.py [hunk 2]
```diff
@@ -92,7 +92,7 @@ def apply_NMT_model(params, args):
         # True when we should score against a reference
 
         #add the test split reference to the dataset
-        path_list = params['DATA_ROOT_PATH'] + s  + '.' + params['PRED_SCORE']
+        path_list = os.path.join(params['DATA_ROOT_PATH'], s + '.' + params['PRED_SCORE'])
```
Comment (drj11) [inline predict.py:95]: I suppose we could consider using `pathlib`: https://stackoverflow.com/a/44189428/242457

File: train.py [hunk 3]
```diff
@@ -16,11 +16,12 @@
 import sys
 
 from timeit import default_timer as timer
+import yaml
 
 from keras_wrapper.extra.read_write import pkl2dict, dict2pkl
 from keras_wrapper.extra.read_write import pkl2dict, dict2pkl
 
-from config import load_parameters
+# from config import load_parameters
```
Comment (drj11) [inline train.py]: delete me

File: train.py [hunk 4]
```diff
@@ -330,8 +329,25 @@ def buildCallbacks(params, model, dataset):
...
     args = parse_args()
     print(args)
-    parameters = load_parameters()
-    if args.config is not None:
+    # load the default config parameters
+    # load the user config and overwrite any defaults
+    if args.config.endswith('.yml'):
+        with open('configs/default-config-BiRNN.yml') as file: #FIXME make this a user option (maybe depend on model type and level?)
+            parameters = yaml.load(file, Loader=yaml.FullLoader)
+        with open(args.config.strip()) as file:
```
Comment (drj11) [inline train.py]: The `.strip()` is weird. And potentially annoying if for some even weirder reason you actually wanted to create a config file with a name that begins with space.

File: train.py [hunk 5]
```diff
@@ -330,8 +329,25 @@ def buildCallbacks(params, model, dataset):
...
+            user_parameters = yaml.load(file, Loader=yaml.FullLoader)
+        for key in user_parameters:
+            parameters[key] = user_parameters[key]
```
Comment (drj11) [inline train.py]: I quite like the clarity of this particular bit of python, but you can also replace it with: `parameters.update(user_parameters)`

File: train.py [hunk 6]
```diff
@@ -351,13 +367,11 @@ def main():
         print ('Error processing arguments: (', k, ",", v, ")")
         exit(2)
```
Comment (drj11) [inline train.py]: As per Guido's main, https://www.artima.com/weblogs/viewpost.jsp?thread=4829, it's better practice to have the function `main()` _return_ its exit status, and have the `if` block that calls main invoke it like: `sys.exit(main())`.
So replace this with `return 2` and replace the call to main, `main()` with `sys.exit(main())`. Or not.

File: train.py [hunk 7]
```diff
@@ -351,13 +367,11 @@ def main():
...
 
-    #check_params(parameters)
-
     check_params(parameters)
-    new_eval_sets=parameters.get('NEW_EVAL_ON_SETS', None)
-    if new_eval_sets != None:
-        set_ar = parameters['NEW_EVAL_ON_SETS'].split(',')
-        parameters['EVAL_ON_SETS'] = set_ar
+    # new_eval_sets=parameters.get('NEW_EVAL_ON_SETS', None)
```
Comment (drj11) [inline train.py]: Delete me

File: train.py [hunk 8]
```diff
@@ -467,9 +481,6 @@ def main():
             train_model(parameters, args.dataset, trainable_est=True, trainable_pred=True, weights_path=parameters.get('PRED_WEIGHTS', None))
 
 
-    elif parameters['MODE'] == 'sampling':
-        logger.error('Depecrated function. For sampling from a trained model, please run predict.py.')
-        exit(2)
```
Comment (drj11) [inline train.py:472]: see earlier comments on Guido's main. replace with `return 2`

File: utils/test-BiRNNdoc.sh [hunk 9]
```diff
@@ -2,26 +2,21 @@
...
 store_path=trained_models/${model_name}/
-patience=5
 level=document
 metric=pearson
 
 # look in utils/testVals.csv for the correct value to compare deepquest test run against
 TESTVAL=$(awk -v backend=$KERAS_BACKEND -v level=$level -v task_name=$task_name -v metric=$metric -F,\
  'NR==1 {next};$1==backend && $2==level && $3==task_name && $4==metric {M=$5};END {print M}' utils/testVals.csv )
 
-rm -rf config.*
```
Comment (drj11) [inline utils/test-BiRNNdoc.sh:19]: hurrah!

**Review by drj11 — APPROVED** [review]
Looks good (assuming it works, I haven't run it). Nothing major.

File: config-doc-BiRNN.yml [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+# SRC_LAN or TRG_LAN are expected to be the file extensions of the data files
+
+# Data files
+TEXT_FILES:
+  'train': 'train.'
+  'val': 'dev.'
+  'test': 'test.'
+
+# Dataset class parameters
+INPUTS_IDS_DATASET: ['source_text', 'target_text']     # Corresponding inputs of the dataset
```
Comment (fredblain) [inline config-doc-BiRNN.yml]: Those variables link the dataset in input, to the inputs of the built model. They are therefore model dependent,
If they are modified or swapped, the model most likely won't work, therefore I would rather have them out of the config file. Similarly, we should probably have a mechanism that checks that the number of input datasets corresponds to the number of input of the model.

**Comment by davidwyld** [comment]
@fredblain as I said on slack:
>I'd like to go ahead and merge it so I can carry on adding some other functions on top of that. I'll merge it into my packaging branch for now, but we can make changes to the config structure if you want, before I merge into the deepquest-dev branch

**Review by fredblain — COMMENTED** [review]
We discussed last Friday that we could have the detault/common variables define under `/utils` rather than having a `configs/default-config-BiRNN.yml` file (which could mislead users about where to define the variables of their experiments, IMO).

Also, I'm not a big fan of all this upper-casing, but unfortunately, keys in python dicts are case-sensitive... I guess it would be ugly to just uppercase the keys when loading the file..!?

**Pull request merged**